File: opentracker_exporter.py
#!/usr/bin/env python3
import os
import requests
import time
from lxml import etree
from prometheus_client import start_http_server, Counter, Gauge


# currying in python looks … wrong (but needed for map)
def parse_attrlist_item(attribute):
    def parse_attrlist_item_item(item):
        key = item.xpath(f'@{attribute}')[0]
        value = int(item.text)
        return key, value
    return parse_attrlist_item_item


class OpentrackerStats:
    def __init__(self, url):
        self.url = f'http://{url}/stats?mode=everything'
        self.update()

    def update(self):
        self.data = etree.fromstring(
            requests.get(self.url).text.encode('utf-8')
        )

        # basic stats
        self.uptime    = self.__get_value('uptime')
        self.torrents  = self.__get_value('torrents/count_mutex')
        self.peers     = self.__get_value('peers/count')
        self.seeds     = self.__get_value('seeds/count')
        self.completed = self.__get_value('completed/count')

        # connection stats
        self.connections = {
            'tcp': {
                'accept':   self.__get_value('connections/tcp/accept'),
                'announce': self.__get_value('connections/tcp/announce'),
                'scrape':   self.__get_value('connections/tcp/scrape')
            },
            'udp': {
                'overall':   self.__get_value('connections/udp/overall'),
                'connect':   self.__get_value('connections/udp/connect'),
                'announce':  self.__get_value('connections/udp/announce'),
                'scrape':    self.__get_value('connections/udp/scrape'),
                'missmatch': self.__get_value('connections/udp/missmatch')
            },
            'livesync': self.__get_value('connections/livesync/count')
        }

        # “debug” stats
        self.renew_count = dict(
            [(int(interval), count) for interval, count in map(
                parse_attrlist_item('interval'),
                self.data.xpath('debug/renew/count')
            )]
        )

        self.http_error = dict(
            map(
                parse_attrlist_item('code'),
                self.data.xpath('debug/http_error/count')
            )
        )

        self.mutex_stall = self.__get_value('debug/mutex_stall/count')

    def __get_value(self, expression):
        return int(self.data.xpath(f'{expression}/text()')[0])


try:
    refresh_interval = os.environ['OPENTRACKER_REFRESH']
except KeyError:
    refresh_interval = 5

tracker_url = os.environ['OPENTRACKER_URL']
stats = OpentrackerStats(tracker_url)

uptime    = Counter('tracker_uptime', 'Second the tracker has been up')
torrents  = Gauge('tracker_torrents', 'Number of tracked torrents')
peers     = Gauge('tracker_peers', 'Number of known peers')
seeds     = Gauge('tracker_seeds', 'Number of known seeds')
completed = Gauge('tracker_completed', 'Number of completed torrents')

connections          = Gauge('tracker_connections', 'Number of connections', ['protocol', 'type'])
connections_livesync = Gauge('tracker_connections_livesync', 'Number of livesync connections')

renew_count = Gauge('tracker_renew_count', 'Number of clients renewing the connection at a specific interval', ['interval'])
http_error  = Counter('tracker_http_error', 'Number of http errors', ['code'])
mutex_stall = Counter('tracker_mutex_stall', '')

start_http_server(9574)

while True:
    # basic stats
    uptime.inc(stats.uptime - uptime._value.get()) # ugly
    torrents.set(stats.torrents)
    peers.set(stats.peers)
    seeds.set(stats.seeds)
    completed.set(stats.completed)

    # connection stats
    for protocol, types in stats.connections.items():
        if protocol != 'livesync':
            for type_, count in types.items():
                connections.labels(protocol, type_).set(count)

    connections_livesync = stats.connections['livesync']

    # “debug” stats
    for interval, count in stats.renew_count.items():
        renew_count.labels(interval).set(count)

    for code, count in stats.http_error.items():
        http_error.labels(code).inc(count - http_error.labels(code)._value.get()) # also ugly

    mutex_stall.inc(stats.mutex_stall - mutex_stall._value.get()) # ugly too

    time.sleep(refresh_interval)
    stats.update()

File: opentracker_exporter.go
package main

import (
	"encoding/xml"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strconv"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	// basic stats
	trackerUptime = prometheus.NewDesc(
		"tracker_uptime_total",
		"Second the tracker has been up",
		nil, nil,
	)
	trackerTorrents = prometheus.NewDesc(
		"tracker_torrents",
		"Number of tracked torrents",
		nil, nil,
	)
	trackerPeers = prometheus.NewDesc(
		"tracker_peers",
		"Number of known peers",
		nil, nil,
	)
	trackerSeeds = prometheus.NewDesc(
		"tracker_seeds",
		"Number of known seeds",
		nil, nil,
	)
	trackerCompleted = prometheus.NewDesc(
		"tracker_completed_total",
		"Number of completed torrents",
		nil, nil,
	)
	trackerMutexStall = prometheus.NewDesc(
		"tracker_mutex_stall_total",
		"",
		nil, nil,
	)

	// connection stats
	trackerConnections = prometheus.NewDesc(
		"tracker_connections_total",
		"Number of clients renewing the connection at a specific interval",
		[]string{"protocol", "type"}, nil,
	)
	trackerConnectionsLivesync = prometheus.NewDesc(
		"tracker_connections_livesync_total",
		"Number of livesync connections",
		nil, nil,
	)

	// debug stats
	trackerRenew = prometheus.NewDesc(
		"tracker_renew_total",
		"Number of renews at a specific interval",
		[]string{"interval"}, nil,
	)
	trackerHTTPError = prometheus.NewDesc(
		"tracker_http_error_total",
		"Number of http errors",
		[]string{"code"}, nil,
	)
)

type TCPConnections struct {
	Accept   float64 `xml:"accept"`
	Announce float64 `xml:"announce"`
	Scrape   float64 `xml:"scrape"`
}

type UDPConnections struct {
	Overall   float64 `xml:"overall"`
	Connect   float64 `xml:"connect"`
	Announce  float64 `xml:"announce"`
	Scrape    float64 `xml:"scrape"`
	Missmatch float64 `xml:"missmatch"`
}

type Connections struct {
	TCP      TCPConnections `xml:"tcp"`
	UDP      UDPConnections `xml:"udp"`
	Livesync float64        `xml:"livesync>count"`
}

type Renew struct {
	Interval int     `xml:"interval,attr"`
	Count    float64 `xml:",chardata"`
}

type HTTPError struct {
	Code  string  `xml:"code,attr"`
	Count float64 `xml:",chardata"`
}

type Stats struct {
	Uptime      float64     `xml:"uptime"`
	Torrents    float64     `xml:"torrents>count_mutex"`
	Peers       float64     `xml:"peers>count"`
	Seeds       float64     `xml:"seeds>count"`
	Completed   float64     `xml:"completed>count"`
	Connections Connections `xml:"connections"`
	Renew       []Renew     `xml:"debug>renew>count"`
	HTTPErrors  []HTTPError `xml:"debug>http_error>count"`
	MutexStall  float64     `xml:"debug>mutex_stall>count"`
}

type Exporter struct {
	URL string
}

func (e Exporter) Describe(ch chan<- *prometheus.Desc) {
	ch <- trackerUptime
	ch <- trackerTorrents
	ch <- trackerPeers
	ch <- trackerSeeds
	ch <- trackerCompleted
	ch <- trackerMutexStall

	ch <- trackerConnections
	ch <- trackerConnectionsLivesync

	ch <- trackerRenew
	ch <- trackerHTTPError
}

func (e Exporter) Collect(ch chan<- prometheus.Metric) {
	resp, err := http.Get(fmt.Sprintf("http://%s/stats?mode=everything", e.URL))
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatal(err)
	}

	stats := Stats{}
	err = xml.Unmarshal(body, &stats)
	if err != nil {
		log.Fatal(err)
	}

	// basic stats
	ch <- prometheus.MustNewConstMetric(trackerUptime, prometheus.CounterValue, stats.Uptime)
	ch <- prometheus.MustNewConstMetric(trackerTorrents, prometheus.GaugeValue, stats.Torrents)
	ch <- prometheus.MustNewConstMetric(trackerPeers, prometheus.GaugeValue, stats.Peers)
	ch <- prometheus.MustNewConstMetric(trackerSeeds, prometheus.GaugeValue, stats.Seeds)
	ch <- prometheus.MustNewConstMetric(trackerCompleted, prometheus.CounterValue, stats.Completed)
	ch <- prometheus.MustNewConstMetric(trackerMutexStall, prometheus.CounterValue, stats.MutexStall)

	// connection stats
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.TCP.Accept, "tcp", "accept")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.TCP.Announce, "tcp", "announce")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.TCP.Scrape, "tcp", "scrape")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.UDP.Overall, "udp", "overall")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.UDP.Connect, "udp", "connect")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.UDP.Announce, "udp", "announce")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.UDP.Scrape, "udp", "scrape")
	ch <- prometheus.MustNewConstMetric(trackerConnections, prometheus.CounterValue, stats.Connections.UDP.Missmatch, "udp", "missmatch")
	ch <- prometheus.MustNewConstMetric(trackerConnectionsLivesync, prometheus.CounterValue, stats.Connections.Livesync)

	// debug stats
	for _, renew := range stats.Renew {
		ch <- prometheus.MustNewConstMetric(trackerRenew, prometheus.CounterValue, renew.Count, strconv.Itoa(renew.Interval))
	}

	for _, httpError := range stats.HTTPErrors {
		ch <- prometheus.MustNewConstMetric(trackerHTTPError, prometheus.CounterValue, httpError.Count, httpError.Code)
	}
}

func main() {
	e := Exporter{URL: os.Getenv("OPENTRACKER_URL")}
	if e.URL == "" {
		log.Fatal("Please specify the environment variable OPENTRACKER_URL")
	}
	prometheus.MustRegister(e)
	http.Handle("/metrics", promhttp.Handler())
	log.Fatal(http.ListenAndServe(":9574", nil))
}
